File: typescript/2022/day4/main.ts
import * as fs from "fs";

const part1: number = fs
  .readFileSync("typescript/2022/day4/test.txt", "utf-8")
  .split("\r\n")
  .map((item: string) => item.split(","))
  .reduce((total: number, pair: string[]) => {
    const pair1 = pair[0].split("-");
    const pair2 = pair[1].split("-");
    const first = _.range(Number(pair1[0]), Number(pair1[1]) + 1);
    const second = _.range(Number(pair2[0]), Number(pair2[1]) + 1);
    if (
      first.every((x: number) => second.includes(x)) ||
      second.every((y: number) => first.includes(y))
    ) {
      return total + 1;
    }
    return total;
  }, 0);

const part2: number = fs
  .readFileSync("typescript/2022/day4/test.txt", "utf-8")
  .split("\r\n")
  .map((item: string) => item.split(","))
  .reduce((total: number, pair: string[]) => {
    const pair1 = pair[0].split("-");
    const pair2 = pair[1].split("-");
    const first = _.range(Number(pair1[0]), Number(pair1[1]) + 1);
    const second = _.range(Number(pair2[0]), Number(pair2[1]) + 1);
    if (first.filter((x: string) => second.includes(x)).length !== 0) {
      return total + 1;
    }
    return total;
  }, 0);

part1;
part2;

File: python/2022/day4/main.py
# Part 1
print(len([0 for pair1, pair2 in [(pair1.split("-"), pair2.split("-")) for pair1, pair2 in [line.strip().split(",") for line in open("python/2022/day4/test.txt")]] if (set([x for x in range(int(pair1[0]), int(pair1[1]) + 1)]).issubset(set([y for y in range(int(pair2[0]), int(pair2[1]) + 1)]))) or (set([x for x in range(int(pair2[0]), int(pair2[1]) + 1)]).issubset(set([y for y in range(int(pair1[0]), int(pair1[1]) + 1)])))]))

# Part 2
print(len([0 for pair1, pair2 in [(pair1.split("-"), pair2.split("-")) for pair1, pair2 in [line.strip().split(",") for line in open("python/2022/day4/test.txt")]] if (set([x for x in range(int(pair1[0]), int(pair1[1]) + 1)]) & (set([y for y in range(int(pair2[0]), int(pair2[1]) + 1)])))]))
